src/editor/Collaboration.js

```diff
@@ -74,40 +101,67 @@ define(function (require, exports, module) {
             var rootDir = StartupState.project("root");
             if(added) {
                 added.forEach(function(addedFile) {
-                    var relPath = Path.relative(StartupState.project("root"), addedFile._path);
-                    // send file only if this client added this file, and not received it
-                    if(_received[relPath]) {
-                        // Clear _received of file name for future events.
-                        delete _received[relPath];
-                        return;
-                    }
-                    FilerUtils.readFileAsBinary(addedFile._path, function(err, buffer) {
-                        if(err) {
-                            console.log(err);
-                        }
-                        var file = new File([buffer], relPath);
-                        _webrtc.getPeers().forEach(function(peer) {
-                            peer.sendFile(file);
-                        });
-                    });
+                    sendFileViaWebRTC(addedFile);
                 });
             }
             if(removed) {
                 removed.forEach(function(removedFile) {
-                    _webrtc.sendToAll("file-removed", {path: Path.relative(rootDir, removedFile.fullPath), isFolder: removedFile.isDirectory});
+                    var relPath = Path.relative(rootDir, removedFile.fullPath);
+                    if(_deletedRemotely[removedFile.fullPath]) {
+                        delete _deletedRemotely[removedFile.fullPath];
+                        return;
+                    }
+                    _webrtc.sendToAll("file-removed", {path: relPath, isFolder: removedFile.isDirectory});
                 });
             }
         });
     };
 
+    function _deleteStructure(found) {
+        var result = new $.Deferred();
+        if(found.length === 0) {
+            result.resolve();
+            return result.promise();
+        }
+
+        var fullPath = found[0];
+        found.splice(0, 1);
+
+        //skip root directory
+        if(fullPath === StartupState.project("root") + '/') {
+            return _deleteStructure(found);
+        }
+
+        console.log("path deleted is " + fullPath);
+        if(fullPath.endsWith('/')) {
```

Comment:
change this function to:
```js
if(found.length === 0) {
    return (new $.Deferred()).resolve().promise();
}

var fullPath = found.shift();

return _deleteStructure(found)
.then(function() {
    if(fullPath === StartupState.project("root") + '/') {
        return;
    }

    return _removeFile(fullPath, fullPath.endsWith('/'));
});
```

Then change `_removeFile` to be promise based instead of callback based.
```js
function _removeFile(fullPath, isFolder) {
    var result = new $.Deferred();
    _deletedRemotely[fullPath] = true;
    var fnName = isFolder ? "getDirectoryForPath" : "getFileForPath";

    FileSystem[fnName](fullPath).unlink(function(err) {
        if(err) {
            result.reject(err);
        } else {
            result.resolve();
        }
    });

    return result.promise();
}
```